### ISSUE TITLE
(z80asm) Parallel build issues

### ISSUE THREAD

**Issue by suborb** [z88dk/z88dk#2422]
The continual "speed up the build" game, gave this this morning:

```
...
z88dk-z80asm -mez80_z80 -I/Users/dom/z88dk/libsrc//../lib/ -Ialgorithm/AStarSearch  -D__CLASSIC -o=algorithm/AStarSearch/obj/ez80_z80/astar_WalkPath.o algorithm/AStarSearch/astar_WalkPath.asm
z88dk-z80asm -mez80_z80 -I/Users/dom/z88dk/libsrc//../lib/ -Ialgorithm/AStarSearch  -D__CLASSIC -o=algorithm/AStarSearch/obj/ez80_z80/astar_WalkPath_callee.o algorithm/AStarSearch/astar_WalkPath_callee.asm
z88dk-z80asm -mkc160 -I/Users/dom/z88dk/libsrc//../lib/ -Ialgorithm/AStarSearch  -D__CLASSIC -o=algorithm/AStarSearch/obj/kc160/astar_CleanUp.o algorithm/AStarSearch/astar_CleanUp.asm
z88dk-z80asm -mkc160 -I/Users/dom/z88dk/libsrc//../lib/ -Ialgorithm/AStarSearch  -D__CLASSIC -o=algorithm/AStarSearch/obj/kc160/astar_DeletePath.o algorithm/AStarSearch/astar_DeletePath.asm
z88dk-z80asm -mkc160 -I/Users/dom/z88dk/libsrc//../lib/ -Ialgorithm/AStarSearch  -D__CLASSIC -o=algorithm/AStarSearch/obj/kc160/astar_PathLength.o algorithm/AStarSearch/astar_PathLength.asm
error: file rename: algorithm/AStarSearch/astar_DeletePath.o~
algorithm/AStarSearch/astar_DeletePath.o~: No such file or directory
error: file open: algorithm/AStarSearch/astar_DeletePath.o
algorithm/AStarSearch/astar_DeletePath.o: No such file or directory
error: file rename: algorithm/AStarSearch/astar_PathLength.o~
algorithm/AStarSearch/astar_PathLength.o~: No such file or directory
error: file rename: algorithm/AStarSearch/astar_SearchResume.o~
algorithm/AStarSearch/astar_SearchResume.o~: No such file or directory
errorerror: : file renamefile open: : algorithm/AStarSearch/astar_CleanUp.o~algorithm/AStarSearch/astar_PathLength.o

algorithm/AStarSearch/astar_PathLength.o: No such file or directory
algorithm/AStarSearch/astar_CleanUp.o~: No such file or directory
make: *** [algorithm/AStarSearch/obj/kc160/astar_PathLength.o] Error 1
make: *** Waiting for unfinished jobs....
make: *** [algorithm/AStarSearch/obj/kc160/astar_DeletePath.o] Error 1
make: *** [algorithm/AStarSearch/obj/ez80_z80/astar_SearchResume.o] Error 1
make: *** [algorithm/AStarSearch/obj/ez80_z80/astar_PathLength.o] Error 1
make: *** [algorithm/AStarSearch/obj/ez80_z80/astar_DeletePath.o] Error 1
make: *** [algorithm/AStarSearch/obj/ez80_z80/astar_CleanUp.o] Error 1
```

Which seems to indicate that the object file is being created adjacent to the .asm file and then moved into location?

**Comment by pauloscustodio** [comment]
Yes, z80asm creates ```target_dir/object_file.o~``` and then renames it to ```target_dir/object_file.o```. This can create a race condition if two processes are building the same file.

**Comment by suborb** [comment]
Ah, because I've not specified a target dir it uses the source directory? So it ends up `mv algorithm/AStarSearch/astar_PathLength.o~algorithm/AStarSearch/obj/jc180/astar_PathLength.o`?

**Comment by pauloscustodio** [comment]
Yes, if you don't specify a target directory, it uses the source directory.

**Comment by pauloscustodio** [comment]
Maybe we could create a lock file in the same directory as the asm file, delaying a second process starting to assemble the same file. I'm just not sure it would work in all platforms, especially Windows,

**Comment by suborb** [comment]
Bingo - it's writing the object file twice!

```
% z88dk-z80asm -v -IXIY -I/Users/dom/z88dk/libsrc//../lib/ -Ialgorithm/AStarSearch -D__CLASSIC -o=algorithm/AStarSearch/obj/ixiy/astar_WalkPath.o algorithm/AStarSearch/astar_WalkPath.asm
Predefined constant: __CLASSIC = 1
Predefined constant: __CPU_Z80__ = 1
Predefined constant: __CPU_ZILOG__ = 1
Predefined constant: __SWAP_IX_IY__ = 1
Predefined constant: __FLOAT_GENMATH__ = 1
Library 'z88dk-z80asm.lib' not found
Reading library '/Users/dom/z88dk/lib/z88dk-z80asm.lib'
Assembling 'algorithm/AStarSearch/astar_WalkPath.asm'
Writing object file 'algorithm/AStarSearch/astar_WalkPath.o'

Writing object file 'algorithm/AStarSearch/obj/ixiy/astar_WalkPath.o'
```

**Comment by pauloscustodio** [comment]
Root cause found: -oobject_file was done to create a consolidated object file, i.e. a linked file with all cross-references between the input files solved (request #999); it assembles all input files, and then links them writing the result -ofile.

Easy solution: if there is only one assembly file input, -o is handled as the output file name of the assembler.

**Comment by suborb** [comment]
Fantastic, thanks @pauloscustodio 

Now I'm getting a race condition on creating the output directory - can we just silently ignore it mkdir fails and it is a directory?

```
z88dk-z80asm -I../ -I/home/dom/z88dk/libsrc//../lib/ -I/home/dom/z88dk/libsrc//../libsrc/ -mez80_z80 -D__CLASSIC  -Oobj/ez80_z80 c/sccz80/qfloat.asm
z88dk-z80asm -I../ -I/home/dom/z88dk/libsrc//../lib/ -I/home/dom/z88dk/libsrc//../libsrc/ -mez80_z80 -D__CLASSIC  -Oobj/ez80_z80 c/sccz80/ftoe.asm
obj/ez80_z80/c: File exists
make: *** [../../Make.config:77: obj/ez80_z80/c/sccz80/ftoe.o] Error 1
```
